KafkaIngester/src/main/java/io/deephaven/kafka/ingest/KafkaIngester.java

```diff
@@ -0,0 +1,298 @@
+    }
+
+    /**
+     * Starts a consumer thread which replicates the consumed Kafka messages to Deephaven.
+     * <p>
+     * This method must not be called more than once on an ingester instance.
+     */
+    public void start() {
+        final Thread t = new Thread(this::consumerLoop, this.toString());
+        t.setDaemon(true);
+        t.start();
+    }
+
+    private static double msgPerSec(final long msgs, final long nanos) {
+        return 1000.0 * 1000.0 * 1000.0 * msgs / nanos;
+    }
+
+    private void consumerLoop() {
+        final long reportIntervalNanos = REPORT_INTERVAL_MS * 1_000_000L;
+        long lastReportNanos = System.nanoTime();
+        final int expectedPartitionCount = openPartitions.size();
+        final DecimalFormat rateFormat = new DecimalFormat("#.0000");
+        while (openPartitions.size() == expectedPartitionCount) {
+            final long beforePoll = System.nanoTime();
+            final long nextReport = lastReportNanos + reportIntervalNanos;
+            final long remainingNanos = beforePoll > nextReport ? 0 : (nextReport - beforePoll);
+            boolean noMore = pollOnce(Duration.ofNanos(remainingNanos));
+            if (noMore) {
+                break;
+            }
+            final long afterPoll = System.nanoTime();
+            if (afterPoll > nextReport) {
+                final long intervalMessages = messagesProcessed - lastProcessed;
+                final long intervalNanos = afterPoll - lastReportNanos;
+                log.info().append(logPrefix)
+                        .append("Processed ").append(intervalMessages).append(" in ")
+                        .append(intervalNanos / 1000_000L).append("ms, ")
+                        .append(rateFormat.format(msgPerSec(intervalMessages, intervalNanos))).append(" msgs/sec")
+                        .endl();
+                lastReportNanos = afterPoll;
+                lastProcessed = messagesProcessed;
+            }
+        }
+        log.info().append(logPrefix).append("Closing Kafka consumer.").endl();
+        consumer.close();
+        throw new UncheckedDeephavenException("Kafka stream was closed.");
+    }
+
+    /**
+     *
+     * @param timeout
+     * @return true if we should abort the consumer thread.
+     */
+    private boolean pollOnce(final Duration timeout) {
+        final ConsumerRecords<?, ?> records;
+        try {
+            records = consumer.poll(timeout);
+        } catch (WakeupException we) {
+            return false;
+        }
+        for (final ConsumerRecord<?, ?> record : records) {
+            final int partition = record.partition();
+            final Consumer<ConsumerRecord<?, ?>> consumer = partitionToConsumer.apply(partition);
+            try {
+                consumer.accept(record);
+            } catch (Exception ex) {
+                ++messagesWithErr;
+                log.error().append(logPrefix).append("Exception while processing Kafka message:").append(ex);
```

Comment:
If we hit any errors, maybe we should deliver a failure to the table downstream. Not necessarily worth plumbing at this stage.

Reply:
https://github.com/deephaven/deephaven-core/issues/934